engine/access/ingestion/collection_syncer.go

```diff
@@ -0,0 +1,398 @@
+//
+// No errors expected during normal operations.
+func (s *CollectionSyncer) findMissingCollections(lastFullBlockHeight uint64) ([]*flow.CollectionGuarantee, int, error) {
+	// first block to look up collections at
+	firstBlockHeight := lastFullBlockHeight + 1
+
+	lastFinalizedBlock, err := s.state.Final().Head()
+	if err != nil {
+		return nil, 0, err
+	}
+	// last block to look up collections at
+	lastBlockHeight := lastFinalizedBlock.Height
+
+	var missingCollections []*flow.CollectionGuarantee
+	var incompleteBlocksCount int
+
+	for currBlockHeight := firstBlockHeight; currBlockHeight <= lastBlockHeight; currBlockHeight++ {
+		collections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return nil, 0, ErrMissingCollection
+		}
+
+		if len(collections) == 0 {
+			continue
+		}
+
+		missingCollections = append(missingCollections, collections...)
+		incompleteBlocksCount += 1
+	}
+
+	return missingCollections, incompleteBlocksCount, nil
+}
+
+// findMissingCollectionsAtHeight returns all missing collections for a specific block height.
+//
+// Expected errors during normal operations:
+// - storage.ErrNotFound if no block is found in storage for the given height
+func (s *CollectionSyncer) findMissingCollectionsAtHeight(height uint64) ([]*flow.CollectionGuarantee, error) {
+	block, err := s.blocks.ByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve block by height %d: %w", height, err)
+	}
+
+	var missingCollections []*flow.CollectionGuarantee
+	for _, guarantee := range block.Payload.Guarantees {
+		inStorage, err := s.isCollectionInStorage(guarantee.CollectionID)
+		if err != nil {
+			return nil, err
+		}
+
+		if !inStorage {
+			missingCollections = append(missingCollections, guarantee)
+		}
+	}
+
+	return missingCollections, nil
+}
+
+// isCollectionInStorage checks whether the given collection is present in local storage.
+// No errors expected during normal operation
+func (s *CollectionSyncer) isCollectionInStorage(collectionID flow.Identifier) (bool, error) {
+	_, err := s.collections.LightByID(collectionID)
+	if err == nil {
+		return true, nil
+	}
+
+	if errors.Is(err, storage.ErrNotFound) {
+		return false, nil
+	}
+
+	return false, fmt.Errorf("failed to retrieve collection %s: %w", collectionID.String(), err)
+}
+
+// RequestCollectionsForBlock conditionally requests missing collections for a specific block height,
+// skipping requests if the block is already below the known full block height.
+func (s *CollectionSyncer) RequestCollectionsForBlock(height uint64, missingCollections []*flow.CollectionGuarantee) {
+	// skip requesting collections, if this block is below the last full block height.
+	// this means that either we have already received these collections, or the block
+	// may contain unverifiable guarantees (in case this node has just joined the network)
+	if height <= s.lastFullBlockHeight.Value() {
+		s.logger.Info().
```

Comment:
let's downgrade this to debug
```suggestion
		s.logger.Debug().
```